snuba/cogs/accountant.py

```diff
@@ -0,0 +1,250 @@
+    program and keep reusing that till you exit.
+
+    Resources are identified as `resource_id` and feature as
+    `app_feature` in the methods. These cannot be enum otherwise we would
+    have to re-release the api every time we add one. And we do not
+    have a single source of truth for them anyway.
+    """
+
+    def __init__(
+        self,
+        granularity_sec: int = 60,
+        topic_name: str = DEFAULT_TOPIC_NAME,
+        queue_size: int = DEFAULT_QUEUE_SIZE,
+        kafka_config: Optional[KafkaConfig] = None,
+        producer: Optional[Producer[KafkaPayload]] = None,
+    ) -> None:
+        """
+        Initializes the accumulator. Instances should be kept around
+        as much as possible as they preserve the instance of the
+        Kafka producers, with it the connections.
+
+        `granularity_sec` defines how often the accumulator will be
+        flushed.
+        """
+        self.__first_timestamp: Optional[float] = None
+        self.__usage_batch: MutableMapping[UsageKey, int] = {}
+        self.__granularity_sec = granularity_sec
+
+        self.__topic = Topic(topic_name)
+
+        if producer is not None:
+            assert kafka_config is None, (
+                "If producer is provided, initialization "
+                "parameters cannot be provided"
+            )
+            self.__producer: Producer[KafkaPayload] = producer
+
+        else:
+            assert kafka_config is not None, (
+                "If no producer is provided, initialization parameters "
+                "have to be provided"
+            )
+
+            self.__producer = KafkaProducer(
+                build_kafka_configuration(
+                    default_config=kafka_config.config_params,
+                    bootstrap_servers=kafka_config.bootstrap_servers,
+                )
+            )
+
+        self.__queue_size = queue_size
+        self.__futures: Deque[Future[BrokerValue[KafkaPayload]]] = deque()
+
+        self.__last_log: Optional[float] = None
+
+    def record(
+        self,
+        resource_id: str,
+        app_feature: str,
+        amount: int,
+        usage_type: UsageUnit,
+    ) -> None:
+        """
+        Record a chunk of usage of a resource for a feature.
+
+        This method is not a blocking one and it is cheap on
+        the main thread. So feel free to call it often.
+        Specifically it takes the system time to associate to
+        the usage to a time range.
+
+        `resource_id` identifies the shared resource.
+          Example: `generic_metrics_indexer_ingestion`
+        `app_feature` identifies the product feature.
+        `amount`  is the amount of resource used.
+        `usage_type` is the unit of measure for `amount`.
+        """
+        now = time.time()
+        if (
+            self.__first_timestamp is not None
+            and now - self.__first_timestamp >= self.__granularity_sec
+        ):
+            self.flush()
```

Comment:
`flush` seems to be a blocking call according to [this](https://github.com/getsentry/snuba/pull/4680/files#diff-2cfdc12931a5cd6ce27d7c7103521551f758ae2c6995cad2dc750989a3ec0a32R176-R177). That would imply that the consumer would be blocked when doing this work. Can the accounting be done from a separate thread instead in order to not impact the throughput of the consumer?

Reply:
The flush call happens at most once every 60s, so the consumer would be blocked once a minute on the flush. If we still want it in it's own thread I can look into it.